pkg/plugins/localfiles.go

```diff
@@ -10,79 +11,189 @@ import (
 	"github.com/grafana/grafana/pkg/util"
 )
 
-var _ fs.FS = (*LocalFS)(nil)
+// CollectFilesFunc is a function that returns a filepath.WalkFunc, which will accumulate its results into acc.
+// The file paths collected into acc should be absolute file paths (like the first argument provided to
+// filepath.WalkFunc)
+type CollectFilesFunc func(acc map[string]struct{}) filepath.WalkFunc
+
+// defaultCollectFilesFunc is a simple CollectFilesFunc that skips all directories and collects all paths into acc.
+var defaultCollectFilesFunc CollectFilesFunc = func(acc map[string]struct{}) filepath.WalkFunc {
+	return func(path string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		acc[path] = struct{}{}
+		return nil
+	}
+}
+
+// emptyCollectFilesFunc is a no-op CollectFilesFunc. It returns a filepath.WalkFunc that does nothing and
+// the accumulator map is not altered in any way. This is suitable to use with plugins.allowListFSNoFiles for tests.
+func emptyCollectFilesFunc(_ map[string]struct{}) filepath.WalkFunc {
+	return func(path string, info fs.FileInfo, err error) error {
+		return nil
+	}
+}
 
 // LocalFS is a plugins.FS that allows accessing files on the local file system.
 type LocalFS struct {
-	// m is a map of relative file paths that can be accessed on the local filesystem.
-	// The path separator must be os-specific.
-	m map[string]*LocalFile
-
-	// basePath is the basePath that will be prepended to all the files (in m map) before accessing them.
+	// basePath is the basePath that will be prepended to all the files to get their absolute path.
 	basePath string
-}
 
-// NewLocalFS returns a new LocalFS that can access the specified files in the specified base path.
-// Both the map keys and basePath should use the os-specific path separator for Open() to work properly.
-func NewLocalFS(m map[string]struct{}, basePath string) LocalFS {
-	pfs := make(map[string]*LocalFile, len(m))
-	for k := range m {
-		pfs[k] = &LocalFile{
-			path: k,
-		}
-	}
+	// collectFilesFunc returns a filepath.WalkFunc that creates a slice of files in this LocalFS, for Files()
+	collectFilesFunc CollectFilesFunc
+}
 
-	return LocalFS{
-		m:        pfs,
-		basePath: basePath,
+// NewLocalFS returns a new LocalFS that can access any file in the specified base path on the filesystem.
+// basePath must use os-specific path separator for Open() to work properly.
+// If collectFilesFunc is not provided, defaultCollectFilesFunc is used instead, which will collect all files.
+func NewLocalFS(basePath string, collectFilesFunc CollectFilesFunc) LocalFS {
```

Comment:
as I understand, specifying `collectFilesFunc` is only useful for tests. To make that clear / easier to use, can you remove the param from this function and expose a different one `NewFakeLocalFS` that receives the collect function?